### ISSUE TITLE
Allow the Deferreds returned by ClientCreator's methods to be cancelled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4329]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4329|
|Type|enhancement|
|Created|2010-03-01 01:54:08Z|
|Branch|https://github.com/twisted/twisted/tree/clientcreator-cancellation-4329|

Cancelling the `Deferred` returned by `ClientCreator`'s `connectTCP`, `connectSSL`, and `connectUNIX` methods should fail the `Deferred` with `CancelledError` and abort the connection attempt by calling `stopConnecting` on the `IConnector` provider returned by the underlying reactor method.

<details><summary>Searchable metadata</summary>

```
trac-id__4329 4329
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_clientcreator_cancellation_4329 branches/clientcreator-cancellation-4329
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1267408448000000 1267408448000000
changetime__1267582758000000 1267582758000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28581]) Branching to 'clientcreator-cancellation-4329'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Done, and with only about 40% more insanity than the RDA.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

 * It would be nice to remove some duplications in ClientCreator, maybe making the canceller a method? Although it may be hard because it relies on the local frame.
 * `test_cancelConnectSSLTimeout` mentions `connectUNIX` in its docstring.

It's going to conflict pretty badly with [#11002](https://github.com/twisted/twisted/issues/11002), because of the changes in proto_helpers, so it would be nice to notify David. Other than that, it looks good, +1!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Oh, and the branch misses a NEWS fragment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [28623]) Refactor ClientCreator methods to use a common implementation

refs [#4329](https://github.com/twisted/twisted/issues/4329)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [28624]) Correct docstring

refs [#4329](https://github.com/twisted/twisted/issues/4329)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [28631]) Merge clientcreator-cancellation-4329

Author: exarkun
Reviewer: therve
Fixes: [#4329](https://github.com/twisted/twisted/issues/4329)

Implement cancellation for the Deferreds returned by `ClientCreator`'s methods.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
